lib/matplotlib/backends/backend_pdf.py

```diff
@@ -1561,6 +1567,28 @@ def _writeImg(self, data, id, smask=None):
         if smask:
             obj['SMask'] = smask
         if mpl.rcParams['pdf.compression']:
+            if data.shape[-1] == 1:
+                data = data.squeeze(axis=-1)
+            img = Image.fromarray(data)
+            img_colors = img.getcolors(maxcolors=256)
+            if colors == 3 and img_colors is not None:
+                # Convert to indexed color if there are 256 colors or fewer
+                # This can significantly reduce the file size
+                num_colors = len(img_colors)
+                img = img.convert(mode='P', dither=Image.NONE,
+                                  palette=Image.ADAPTIVE, colors=num_colors)
+                data, bit_depth, palette = self._writePng(img)
```

Comment:
```suggestion
                png_data, bit_depth, palette = self._writePng(img)
```
For clarity we shouldn't reuse the variable *data* with a different meaning.